Translations v2

daneryl opened the pull request:
fixes #3344 

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

daneryl committed:
WIP, CreateTranslations Service

daneryl committed:
folder structure

daneryl committed:
rename CreateTranslations

daneryl committed:
GetTranslationsService

daneryl committed:
UpdateTranslationsService

daneryl committed:
WIP, integration new translations on old model

daneryl committed:
Translations v2 upsert on update using bulk

daneryl committed:
updateTranslationsV2 feature flag condition

daneryl committed:
DeleteTranslationsService and integration with old api

daneryl committed:
remove unused processSystemKeys function

daneryl committed:
remove unused translations function addEntry

daneryl committed:
addLanguage v2 support

daneryl committed:
tests for DeleteTranslationsService

daneryl committed:
test CreateTranslationsService upsert

daneryl committed:
moved upsert to its own service

daneryl committed:
WIP, v2 migration on get

daneryl committed:
WIP, migrate translations on get and flag active

daneryl committed:
WIP, v2 Get with an _id as query

daneryl committed:
old _id maps to language for the new collection

daneryl committed:
reminder on where to continue

daneryl committed:
fix type errors

daneryl committed:
fix lint

daneryl committed:
Validate language on CreateTranslation

daneryl committed:
validate language on upsertTranslationsService

daneryl committed:
translations save now uses locale to determine update / create

daneryl committed:
move migration to its own, file
the migration should be moved to migrations folder and deploy normally
as a last step

daneryl committed:
migration to its own file

daneryl committed:
fix cognitive complexity

daneryl committed:
tweak specs descriptions

daneryl committed:
"idempotent" migration

LaszloKecskes commented on app/api/i18n.v2/model/Translation.ts:
```diff
@@ -0,0 +1,29 @@
+  readonly language: string;
+
+  readonly context: {
+    type?: 'Entity' | 'Relationship Type' | 'Uwazi UI' | 'Thesaurus';
```
Can a context not have type?

fnocetti replied:
Also, maybe its a good idea to extract the type to it's own type definition.

daneryl replied:
No, i do not think type should be optional, going to change that and extract the type also :+1:

daneryl committed:
Merge branch 'production' into translations_v2

daneryl committed:
tweaks for client
- support for context filtering
- client now tracks translation by locale instead of _id

daneryl committed:
fix preserveSync fixtures

daneryl committed:
fix types

daneryl committed:
add tests to GetTranslationServices

daneryl committed:
Merge remote-tracking branch 'origin/production' into translations_v2

daneryl committed:
WIP, updateContext v2 support

daneryl committed:
WIP, validate translations create has all the languages
refactor validation code
create an addLanguage v2

daneryl committed:
migration, adds missing language keys

daneryl committed:
unique validation test for translation creation

daneryl committed:
WIP, duplicatedKey error

daneryl committed:
throw error when trying to create a duplicated Translation

daneryl committed:
Merge remote-tracking branch 'origin/production' into translations_v2

daneryl committed:
Some tweaks

daneryl committed:
Merge remote-tracking branch 'origin/production' into translations_v2

LaszloKecskes commented on app/api/i18n.v2/schemas/TranslationDBO.ts:
```diff
@@ -0,0 +1,10 @@
+export interface TranslationDBO {
+  language: string; // should be an enum ?
```
Good question. If we have a list of supported languages somewhere with their keys, probably we should make use of that, where we define languages.

daneryl replied:
Added as a String literal union type on the original SettingsSchema

LaszloKecskes commented on app/api/i18n.v2/services/CreateTranslationsService.ts:
```diff
@@ -0,0 +1,72 @@
+import { Translation } from '../model/Translation';
+
+export interface CreateTranslationsData {
+  language: string; // should be an enum ?
```
Same as in the DBO, probably yes.

LaszloKecskes commented on app/api/i18n.v2/database/MongoTranslationsDataSource.ts:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  async deleteByContextId(contextId: string) {
+    return this.getCollection().deleteMany({ 'context.id': contextId });
```
For this, add an index on the collection just for context.id

LaszloKecskes replied:
Let's probably make it a compound with the order (context.id, key), since later on there are queries that use only these two.

daneryl replied:
done :+1:

LaszloKecskes commented on app/api/i18n.v2/database/MongoTranslationsDataSource.ts:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  async deleteByLanguage(language: string) {
+    return this.getCollection().deleteMany({ language });
```
Same: an index for language.

LaszloKecskes replied:
Scratch that, the current one covers this.

LaszloKecskes commented on app/api/i18n.v2/database/MongoTranslationsDataSource.ts:
```diff
@@ -0,0 +1,195 @@
+          t => t
+        );
+
+        for await (const dbt of dbtrans) {
```
With Fede we usually use 
```
while(await cursor.hasNext()) {
    const item = await cursor.next();
```
It's a lot less pretty, but avoids the `/* eslint-disable no-restricted-syntax */` problem. On the downside, it also adds a lot of no-await-in-loops. Maybe we could discuss disabling that one on the tech call, since we seem to use it a lot.

daneryl replied:
changed to a while :+1:

LaszloKecskes commented on app/api/i18n.v2/migrations/index.js:
```diff
@@ -0,0 +1,107 @@
+  if (translation.contexts?.length) {
+    return translation.contexts.reduce(
+      (flatTranslations, context) =>
+        flatTranslations.concat(
```
I think concat copies both arrays, so inside a reduce this can be slow. Alternatives can be pushing one by one inside reduce and always pass the reference to the array, or I think this one might be solved by a douple map, then flat().

daneryl replied:
done :+1:

LaszloKecskes commented on app/api/i18n.v2/services/specs/DeleteTranslationsService.spec.ts:
```diff
@@ -0,0 +1,149 @@
+  );
+
+const fixtures = {
+  translations_v2: [
```
This is a common comment for all tests: in common.v2/testing we've started with experimenting how we could use the fixture factory for v2, so if you'd like, you could add to that next time. For now, I don't think it is necessary.

daneryl replied:
Im going to have to change all fixtures on the final iteration, i will add this change to that development :+1:

LaszloKecskes commented on app/api/i18n/specs/translations_v2_support.spec.ts:
```diff
@@ -0,0 +1,793 @@
+
+  describe('updateContext', () => {
+    describe('when feature flag is on', () => {
+      it('should properly change context name, key names, values for the keys changed and deleteProperties, and create new values as new translations (OMG !)', async () => {
```
Left an extra comment at the end of the text description.

daneryl replied:
removed it :)

LaszloKecskes commented on app/api/i18n/v2_support.ts:
```diff
@@ -0,0 +1,217 @@
+
+const flattenTranslations = (translation: TranslationType): CreateTranslationsData[] => {
+  if (translation.contexts?.length) {
+    return translation.contexts.reduce(
```
Same comment as in the migration: try to go without the concat in reduce.

daneryl replied:
done :+1:

LaszloKecskes commented on app/api/i18n/v2_support.ts:
```diff
@@ -0,0 +1,217 @@
+  }) as EnforcedWithId<TranslationType>[];
+};
+
+export const createTranslationsV2 = async (translation: TranslationType) => {
```
For the next couple, where the dependency tree's are initialized: we usually write out the initialization separately, so they are easier to use. For data sources, see files named data_source_defaults, for services, see files named service_factories. Relationships.v2 is the best example, since that is a lot.

daneryl replied:
done :+1:  thank you, now all transactions.run are proper transactions!

LaszloKecskes commented on app/api/i18n/translations.ts:
```diff
@@ -175,44 +186,64 @@ const update = async (translation: TranslationType | IndexedTranslations) => {
-      })
-    );
-    return 'ok';
+  async save(translation: TranslationType | IndexedTranslations) {
```
So, the idea is that we save to both until we think we are ready to drop the v1?

daneryl replied:
Yes, this gives us the ability to deactivate if something goes wrong to use the old code, while we work on a fix.

LaszloKecskes replied:
Nice!

LaszloKecskes reviewed (changes requested):
There are a couple of local things, but overall now everything is there I think. The online migration is impressive. :)

It's a pity we can't run the entire api unite test suite with the feature flag on to see what happens.

daneryl committed:
push instead of concat

daneryl commented:
> There are a couple of local things, but overall now everything is there I think. The online migration is impressive. :)
> 
> It's a pity we can't run the entire api unite test suite with the feature flag on to see what happens.

Thanks for the review, working on some changes, i did run it locally with the flag activated and everything was green btw, but its a good point maybe we should be able to somehow run it on ci ?

daneryl committed:
change for await for a while loop

daneryl committed:
languagekeys as enum

daneryl committed:
add context.id/key compound index

daneryl committed:
remove comment

daneryl committed:
emit types

daneryl committed:
failing updateContext transaction

daneryl committed:
perform proper transactions reusing transaction manager

daneryl committed:
insert now bubbles up any other error other than duplicate key

daneryl committed:
use session for all db operations

daneryl committed:
translation validation service

daneryl committed:
deleteTranslationsv2 behind feature flag

daneryl committed:
ci cypress prints server output

daneryl committed:
Merge remote-tracking branch 'origin/production' into translations_v2

LaszloKecskes commented on app/api/i18n.v2/contracts/TranslationsDataSource.ts:
```diff
@@ -0,0 +1,25 @@
+
+  updateValue(key: string, contextId: string, language: string, value: string): Promise<void>;
+
+  calculateUnexistentKeys(keys: string[]): Promise<string[]>;
```
Minimal english nitpick, but according to my dictionary it is either: unexisting, nonexistent, or inexistent.

daneryl replied:
:+1: changed to calculateNonexistentKeys

LaszloKecskes commented on app/api/i18n.v2/database/MongoTranslationsDataSource.ts:
```diff
@@ -0,0 +1,150 @@
+    const stream = this.createBulkStream();
+
+    await Promise.all(
+      items.map(async item =>
```
I did not catch this last time, but maybe it's not a good idea to use the stream functions parallel, since the underlying structure holding the information is just a regular array. So while it is most likely extremely rare, there might be some race conditions involved (with multiple things triggering a flush at the same time, etc.).

LaszloKecskes replied:
...and actually, later in the same file (updateKeysByContext) it's synchronized through a promise based reduce. I think something like that would be the good approach.

daneryl replied:
:+1: changed to a reduce

LaszloKecskes commented on app/api/i18n.v2/migrations/index.js:
```diff
@@ -0,0 +1,107 @@
+/* eslint-disable no-await-in-loop */
+const flattenTranslations = (translation, languagesByKeyContext) => {
```
I don't remember exactly, but should we move this to the normal migrations?

daneryl replied:
Will move it on the next iteration, for now this migration runs when activating a feature flag

LaszloKecskes reviewed (changes requested):
Aside from those two small things, looks good.

daneryl committed:
tweak upsert and typo

LaszloKecskes reviewed (approved):
Looks good!
